docs/proposals/scalable-alertmanager.md

```diff
@@ -0,0 +1,78 @@
+
+This mechanism covers multiple challenges (scaling up & down, rolling restarts, total cluster failure). **To top this off, in the implementation, we’ll always try to request state from other replicas in the cluster before ever trying to go to object storage.**
+
+### Replication & Consistency
+
+Upstream Alertmanager replicates notification state between replicas to ensure notifications are not sent more than once. Cortex Alertmanager does the same.  When we move to a model where AM instances for a given tenant only live on a subset of replicas, we have to decide how we will keep these replicas in sync.
+
+**We have an option of doing nothing** and use the existing gossip mechanism and gossip all tenants state to all replicas. The AM Router will then drop state updates for tenants which don’t shard to a given replica. I think this will be easy to implement as it requires  few changes but probably won’t scale.
+
+**However, I propose we Synchronized state over gRPC.** The upstream Alertmanager notification dispatcher uses a timeout-based approach for its notifications. Using the “peer order”, it’ll wait a certain amount of time before letting other replicas know if this notification succeeded or not. If it didn’t, the next replica in line will try to send the notification. We propose to communicate this via gRPC calls.
+
+While I propose the use of gRPC, is good to note that doing nothing is still a solid candidate for consideration but not having to maintain and operate two separate replication patterns feels like a win.
+
+### Architecture
+
+To implement the sharding strategy we need a component in charge of handling incoming alerts and API requests, distributing it to the corresponding shard.
+
+The idea here is to have an “Alertmanager Distributor” as a first stop in the reception of alerts. Once alerts are received, the component is in charge of validating the alerts against the limits. Validated alerts are then sent to multiple managers in parallel.
+
+The individual pieces of this component (sharding, limits) cannot be optional - the optional part of it is where we decide to run it.
+
+We can either run it as a separate service or embed it. **I propose we simply embed it**. At its core it’ll be simpler to operate. With future work making it possible to run as a separate service so that operators can scale when/if needed.
+
+
+## Conclusion
+
+Under the assumption we implement the options proposed above, our architecture looks like this:
+
+![Scalable Alertmanager Architecture](/images/proposals/scalable-am.png)
+
+**POST /api/v1/alerts (from the ruler) can go to any Alertmanager replica.** The AM distributor uses the ring to write alerts to a quorum of AM managers (reusing the existing code). We continue to use the same in-memory data structure from the upstream Alertmanager to save alerts and notify other pieces
```

Comment:
**question:** In this architecture the ruler will still use HTTP to forward alerts to the Alertmanager. In order to avoid multiple deliveries will the Ruler only be configured with a single address? If so does this assume the user will need to place a loadbalancer between the `ruler` and the `am-distributor` to achieve an equal distribution of traffic and ensure availability?

Reply:
I think the short answer is yes. 

Just like today, you need to use a headless service (or include all the alertmanager URLs manually) to avoid load balancing traffic between Alertmanagers. You'll need to ensure you have a way to load balance traffic between the ruler <> Alertmanager.

Reply:
👍